aries/agents/rust/mediator/src/persistence/errors.rs

```diff
@@ -0,0 +1,62 @@
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+#[error("No account found matching given input: {0}")]
+pub struct AccountNotFound(pub String);
+
+/// Error closely related to the storage backend
+#[derive(Error, Debug)]
+#[error(transparent)]
+pub struct StorageBackendError {
+    #[from]
+    pub source: Box<dyn std::error::Error>,
+}
+
+/// Error parsing values from backend into expected structures
+#[derive(Error, Debug)]
+#[error("Couldn't retrieve or decode expected data: {0}")]
+pub struct DecodeError(#[from] pub Box<dyn std::error::Error>);
+
+/// Creates an error enum composed of individual error items given as list.  
+/// The enum variants are named identically to the error variants provided.
+/// From<> impls (to the composition) and Display are automatically derived
+/// with the help of thiserror.
+/// Usage:
+/// errorset!(ComposedError[ErrorVariant1, ErrorVariant2]);
+
+macro_rules! errorset {
+    ($errorset_name:ident[$($error_name: ident),*]) => {
+        #[derive(Error, Debug)]
+        pub enum $errorset_name {
+            $(
+            #[error("{0}")]
+            $error_name(#[from] $error_name),
+            )*
+            /// Generic error variant - display, backtrace passed onto source anyhow::Error
+            /// Useful for chucking in errors from random sources. See usage of anyhow! macro.
+            #[error(transparent)]
+            ZFhOt01Rdb0Error(#[from] anyhow::Error),
+        }
+    };
+}
+
+errorset!(GetAccountIdError[StorageBackendError, AccountNotFound]);
+errorset!(ListAccountsError[StorageBackendError, DecodeError]);
+
+errorset!(GetAccountDetailsError[StorageBackendError, AccountNotFound, DecodeError]);
+errorset!(RetrievePendingMessageCountError[StorageBackendError, AccountNotFound]);
+errorset!(RetrievePendingMessagesError[StorageBackendError, AccountNotFound]);
+errorset!(AddRecipientError[StorageBackendError, AccountNotFound]);
+// Same error modes as AddRecipientError
+pub type RemoveRecipientError = AddRecipientError;
```

Comment:
hmmm, perhaps indication of what can be problematic with this error management pattern

The thing, is, `RetrievePendingMessagesError, RetrievePendingMessageCountError, AddRecipientError, RemoveRecipientError` are all the same set of errors, but it's difficult to give the set proper name. The same set of errors might, as we observe, much possibly be returned by various unrelated functions.
I wanted to suggest - why don't we name `[StorageBackendError, AccountNotFound]` set as `ModificationError`, but it would be confusing to return such named error from `retrieve_pending_message_count`. 
On the other hand, if we define `ModificationError` as `[StorageBackendError, AccountNotFound]`, it's possible some other modifying function would actually want to return different set of errors.
So it seems that finding right name for "sets" is just not sensible goal. 

I am commenting on this because of the asymmetry - given that `RetrievePendingMessageCountError, RetrievePendingMessagesError, AddRecipientError` are unique error types, despite representing same set of circumstances underneath, we should probably keep the approach also with  `RemoveRecipientError` and defined is a new type, rather than alias.
Unless there is argument why to give a different treatment to `RemoveRecipientError` and say `RetrievePendingMessagesError`

Reply:
Also comes to my mind that with this approach of - error type per operation - perhaps it would make sense to defined this errors nearby the function they are being thrown from.

Reply:
> I am commenting on this because of the asymmetry - given that `RetrievePendingMessageCountError, RetrievePendingMessagesError, AddRecipientError` are unique error types, despite representing same set of circumstances underneath, we should probably keep the approach also with `RemoveRecipientError` and defined is a new type, rather than alias. Unless there is argument why to give a different treatment to `RemoveRecipientError` and say `RetrievePendingMessagesError`

One way to think about this is that `RetrievePendingMessagesError`, `RetrievePendingMessageCountError`, `AddRecipientError` are all actually not the same set. They are different sets corresponding to different operations that happen to have the same contents. Each may be individually extended or shrunk as the code/implementation evolves. Hence, they need to be different types.

For example, in future, we could add an error case to `AddRecipientError` ~  `WritesAcessDenied`. Which would not be relevant to the others. 

`RemoveRecipientError`, `AddRecipientError` are slightly different because their functions are performing pretty much the same operation (modification). And are expected to fail similarly.

Reply:
> Also comes to my mind that with this approach of - error type per operation - perhaps it would make sense to defined this errors nearby the function they are being thrown from.

Yes. I was thinking the same. Here, since these composed errors are used in a trait definition (with no default implementation) I choose to keep them all together along with the base errors. To keep things tidy.

Reply:
Fair enough. Let's see how it develops as time goes :-)